### PR TITLE
Dart Plugin Registrant - Reland e97c8016e398

### PR DESCRIPTION
**Pull request by gaaclarke**
The original commit is in this PR.  The difference is in the second commit and instead of hunting for the dart plugin registrant we look for it referenced at a particular place.

Framework PR: https://github.com/flutter/flutter/pull/100572
Relands: https://github.com/flutter/engine/pull/31418
The revert: https://github.com/flutter/engine/pull/32162

Integration test in framework PR.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: runtime/dart_plugin_registrant.cc [hunk 1]
```diff
@@ -65,12 +39,30 @@ bool InvokeDartPluginRegistrantIfAvailable(Dart_Handle library_handle) {
...
+  if (Dart_IsError(library)) {
     return false;
   }
+  Dart_Handle registrant_file_uri =
+      Dart_GetField(library, tonic::ToDart("dartPluginRegistrantLibrary"));
+  if (Dart_IsError(registrant_file_uri)) {
+    // TODO(gaaclarke): Find a way to remove this branch so the field is
+    // required. I couldn't get it working with unit tests.
+    return InvokeDartPluginRegistrantIfAvailable(library);
+  } else {
```
Comment (blasten) [inline runtime/dart_plugin_registrant.cc:56]: nit: else not required

File: runtime/dart_plugin_registrant.cc [hunk 2]
```diff
@@ -65,12 +39,30 @@ bool InvokeDartPluginRegistrantIfAvailable(Dart_Handle library_handle) {
...
+    std::string registrant_file_uri_string =
+        tonic::DartConverter<std::string>::FromDart(registrant_file_uri);
+    if (registrant_file_uri_string.empty()) {
+      FML_LOG(ERROR) << "Unexpected empty dartPluginRegistrantLibrary.";
+      return false;
+    } else {
```
Comment (blasten) [inline runtime/dart_plugin_registrant.cc:62]: ditto

**Review by blasten — APPROVED** [review]

File: runtime/dart_plugin_registrant.cc (hunk 1 above)
Comment (gaaclarke) [reply runtime/dart_plugin_registrant.cc]: done

File: runtime/dart_plugin_registrant.cc (hunk 2 above)
Comment (gaaclarke) [reply runtime/dart_plugin_registrant.cc]: done

**Pull request merged**